PIR to RIR translation

JanJecmen opened the pull request:
This is WIP. The ideas for the first pass are:
- just get something working
- register allocation is the dumbest thing possible (phi and inputs are in the same slot, other that that, everything gets its own slot)
- use `ldloc_` and `stloc_` for every instruction arguments and result (there are some optimizations to this like load/store is copied directly, store/load is removed if the store is not needed for later)
- a lot is still missing, most notably function calls

Also, a lot of it is clang format...

JanJecmen committed:
wip

JanJecmen committed:
simple test

JanJecmen committed:
clang format

JanJecmen committed:
merge master into p2r

JanJecmen committed:
fix renaming Function to Closure

JanJecmen committed:
declare number of slots

JanJecmen committed:
wip arg passing for pir

JanJecmen committed:
disable failing wip test

JanJecmen committed:
reuse ldarg_ for PIR's LdArg

JanJecmen committed:
pir calling convention first attempt

JanJecmen committed:
pir 2 rir tests refactor

JanJecmen committed:
p2r fixes and more tests

JanJecmen committed:
remove incorrect cleanup

JanJecmen committed:
wip p2r tests - control flow

JanJecmen committed:
wip p2r tests - control flow 2

JanJecmen committed:
fix code verifier

JanJecmen committed:
change force_ instruction for PIR use

JanJecmen committed:
fix environment proxy when env is created

JanJecmen committed:
tests for loops in pir

JanJecmen committed:
rir cleanup of ldloc and stloc

JanJecmen committed:
fix comment

JanJecmen committed:
add tracking of source indices through pir + clang format

JanJecmen committed:
fix phi copies insertion

JanJecmen committed:
pir 2 rir add instructions

JanJecmen committed:
fix pir type bug

charig commented on rir/src/compiler/pir/bb.h:
```diff
@@ -67,6 +67,7 @@ class BB {
     Instrs::iterator insert(Instrs::iterator it, Instruction* i);
+    Instrs::iterator insert(Instrs::iterator it, Instrs is);
     void replace(Instrs::iterator it, Instruction* i);
 
```
This last two are a little weird. What is Instrs? Just a list of instructions?

JanJecmen replied:
Yes, it's just a `std::vector<Instruction*>`, but this is a leftover piece of code that is not used currently.. I removed it for now.

JanJecmen committed:
control flow test

JanJecmen committed:
merge master into p2r

JanJecmen committed:
fix merge

o- commented:
is this up for review, or still wip?

JanJecmen commented:
~both, kind of :)~
wip

JanJecmen committed:
add locals count tracking, refactor Pir2Rir

JanJecmen committed:
call test

JanJecmen committed:
merge master into p2r

JanJecmen committed:
tests

JanJecmen committed:
getvar_ and refactoring

JanJecmen committed:
calls first attempt

JanJecmen committed:
Merge branch 'master' into p2r

JanJecmen committed:
fix merge

o- commented:
UPDATE: I might have changed my mind on this. I can see how it makes some things easier...

hey, so first question: is `PirCopy` necessary?

it seems to me that you could do the following instead:
1. create a list of all `Phi`
2. whenever you are done compiling a BB `bb`, you do the following:

```
for (phi : listOfPhis) {
  if (phi->input.find(bb))
    // do whatever Tag::PirCopy: case does in the pir2rir compiler
```

But, probably I am missing something?

o- commented on rir/src/compiler/pir/instruction.h:
```diff
@@ -773,10 +784,21 @@ typedef SEXP (*CCODE)(SEXP, SEXP, SEXP, SEXP);
 
 class VLI(CallBuiltin, Effect::Any, EnvAccess::Write) {
   public:
+    constexpr static size_t callArgOffset = 1;
```
I don't think that's necessary here. Every argument to the builtin instruction are arguments for the builtin.

Call has a callArgOffset, because the first instruction argument is the closure. So the first call argument starts at index 1. but here the builtin is an immediate argument (part of the instruction) and not an ir-level input to the instruction. (like in callEager)

o- commented on rir/src/compiler/pir/instruction_list.h:
```diff
@@ -61,6 +61,7 @@
     V(Force)                                                                   \
-    V(CastType)
+    V(CastType)                                                                \
+    V(PirCopy)
```
see comment bellow, it would be great if we can do it without this instruction. The reason is, I am afraid somebody will use it earlier in the compiler and suddenly we have aliasing to deal with.

o- commented on rir/src/compiler/translations/pir_2_rir.cpp:
```diff
@@ -0,0 +1,601 @@
+                return;
+            }
+            case Tag::MkArg: {
+                break;
```
maybe `assert(uses are all call instruction)`

o- commented on rir/src/compiler/translations/pir_2_rir.cpp:
```diff
@@ -0,0 +1,601 @@
+
+                std::vector<FunIdxT> callArgs;
+                call->eachCallArg([&](Value* arg) {
+                    auto mkarg = MkArg::Cast(arg);
```
can we add a todo that we are ignoring `mkarg->arg<0>()` (the potentially eagerly computed value slot). It's fine to ignore it for now, but we should probably add it at some point later.

o- commented on rir/src/interpreter/interp.cpp:
```diff
@@ -119,6 +119,9 @@ static void jit(SEXP cls, Context* ctx) {
 
 void tryOptimizeClosure(DispatchTable* table, size_t offset, Context* ctx) {
 
+    // TODO: needs to change for PIR
```
hmm, not super happy about dead code. I would suggest we remove this rir2rir optimization stuff and focus on rir2pir2rir optimization.

o- commented on rir/src/ir/insns.h:
```diff
@@ -80,6 +85,11 @@ DEF_INSTR(stvar2_, 1, 1, 0, 1)
+/**
+ * copyloc_:: copy one local into another
+ */
+DEF_INSTR(copyloc_, 2, 0, 0, 1)
```
can we give this another name? `movloc` maybe. I find `copying` a bit a heavy word for the operation.

o- commented on rir/src/optimization/localize.h:
```diff
@@ -36,6 +36,8 @@ class Localizer : public InstructionDispatcher::Receiver {
         auto v = analysis[ins][sym];
 
+// Disable as part of PIR work... ldarg_ reused for pir, since this will go away
+#if 0
```
again, I think we should just merge #82 or something similar, instead  of keeping dead code

o- commented on rir/src/runtime/Function.h:
```diff
@@ -61,6 +61,7 @@ struct Function {
         envChanged = false;
         deopt = false;
         markOpt = false;
+        isPirCompiled = false;
```
is this neccessary? isn't the slot/signature mechanism enough?

o- commented on rir/src/compiler/translations/pir_2_rir.cpp:
```diff
@@ -0,0 +1,601 @@
+            case Tag::StVar: {
+                auto stvar = StVar::Cast(instr);
+                loadEnv(it, stvar->env());
+                cs << BC::setEnv() << BC::stvar(stvar->varName);
```
can we at least keep track of the last setEnv and avoid if unnecessary?

o- commented on rir/src/compiler/translations/pir_2_rir.cpp:
```diff
@@ -0,0 +1,601 @@
+                break;
+            }
+            case Tag::CastType: {
+                // TODO: what should it do?
```
nop should be fine

o- commented on rir/src/compiler/translations/pir_2_rir.cpp:
```diff
@@ -0,0 +1,601 @@
+            }
+            case Tag::CallBuiltin: {
+                auto blt = CallBuiltin::Cast(instr);
+                // TODO: is it needed to set the environment?
```
i would think so

o- commented on rir/src/compiler/translations/pir_2_rir.cpp:
```diff
@@ -0,0 +1,601 @@
+                auto mkenv = MkEnv::Cast(instr);
+                loadEnv(it, mkenv->parent());
+                cs << BC::makeEnv() << BC::dup() << BC::setEnv();
+                // bind all args
```
maybe we should have an instruction that does this in one go? not for now, but in the future...

o- commented on rir/src/compiler/translations/pir_2_rir.cpp:
```diff
@@ -0,0 +1,601 @@
+            }
+            case Tag::Phi: {
+                auto phi = Phi::Cast(instr);
+                load(it, phi->arg(0).val());
```
I don't understand this. Why can you take the input 0? because you know that all of the inputs are in the same local value?

o- reviewed (commented):
yeah, finally we can run stuff. did a first pass. thanks for the PR.

o- commented:
small note: can you squash the commits? Eg.:

```
git checkout master
git checkout -b p2r-squash
git merge --squash p2r
git add rir && git commit
```

JanJecmen committed:
cleanup

JanJecmen committed:
cleanup

JanJecmen committed:
fix monomorphic site speculation branch targets

JanJecmen committed:
mkarg

JanJecmen committed:
address some pr comments

mhyee commented on rir/src/compiler/util/visitor.h:
```diff
@@ -199,7 +199,7 @@ class Visitor
     : public VisitorImplementation<true, VisitorHelpers::IDMarker> {};
 class UnstableIDsVisitor
-    : public VisitorImplementation<false, VisitorHelpers::PointerMarker> {};
+    : public VisitorImplementation<true, VisitorHelpers::PointerMarker> {};
```
Why is this changed to stable ordering?

JanJecmen replied:
Oh, this was when I was playing with the reg. allocation.. It's used only for renumbering the BBs after optimizations, and I wanted to have the IDs stable. Not needed now I guess.

o- replied:
uhm, this should probably be stable, since, as you say it's used for renumbering... this looks like a bug, gonna look into it

JanJecmen committed:
track setting env

JanJecmen committed:
merge master

JanJecmen committed:
merge master

mhyee commented on rir/src/api.cpp:
```diff
@@ -114,17 +116,58 @@ REXPORT SEXP rir_analysis_liveness(SEXP what) {
+    auto fun = p2r(m, oldFun);
+    p(fun->container());
+
+    // TODO: put instead into a new table slot...
```
I think we need to figure out exactly what we want to do here. I was under the impression that in this case, the new version should be in the same table slot (because the calling convention or interface is the same), but in the linked list of versions. In other words, we use a different slot if the caller needs to do something different.

But part of this reasoning was because we had the old optimization/deoptimization system, which has been removed (#82). So maybe we should rethink this.

JanJecmen replied:
I agree we should think about it. For PIR, the caller indeed needs to do something different, namely it should not allocate an environment for the callee. I am working now on supporting the slots mechanism.

o- replied:
> I think we need to figure out exactly what we want to do here. I was under the impression that in this case, the new version should be in the same table slot (because the calling convention or interface is the same)

hmm, ok, I guess there are two ways to look at it. you could say, that in both cases the env is set up by the callee. just in pir optimized it is explicit and in not pir optimized it's implicit.

but I find the more natural way is to say we have two cc, one is: "env is caller setup", thats what R and unoptimized rir does. And then we have "env is callee setup", which is what optimized rir does.

mhyee commented on rir/src/api.cpp:
```diff
@@ -114,17 +116,58 @@ REXPORT SEXP rir_analysis_liveness(SEXP what) {
+
+    // TODO: put instead into a new table slot...
+
+    // patch the closure
```
I think this is just asking for a method that inserts a new version into an existing slot and updates all the pointers/fields.

JanJecmen replied:
Actually, I am not sure about these... 

- `envChanged` and `envLeaked` were used for rir optimizations that went away, right @o-? Should we still keep them?
- `invocationCount` for a new version should start from 0, I guess?
- `signature`: I think this still needs discussion, maybe new slot = different signature?

o- replied:
> Should we still keep them?

no

> invocationCount for a new version should start from 0, I guess?

do we count per version or per function? if the former, then probably 0, yeah

mhyee replied:
> I think this still needs discussion, maybe new slot = different signature?

I think this was the original plan. Each slot is described by a signature, if two versions have the same signature then they have the same slot.

mhyee commented on rir/src/compiler/translations/pir_2_rir.cpp:
```diff
@@ -0,0 +1,601 @@
+                phi->eachArg([&](Value* arg) {
+                    // here the arg must not yet be allocated
+                    if (shouldAllocate(arg))
+                        allocateLocal(arg, slot);
```
I don't understand, why is the same slot being used for all the phi's arguments?

JanJecmen replied:
That is the out-of-SSA transition: you remove the `Phi` by assigning it to a register, and making sure that all inputs are in that same register. Then when executing, the `Phi` is a `noop` because you know that regardless of which path you took, the value is in the `Phi`'s register.

mhyee commented on rir/src/compiler/translations/pir_2_rir.cpp:
```diff
@@ -0,0 +1,601 @@
+        push(ast);
+    }
+
+    FunIdxT pop(size_t localsCnt) {
```
Hmm, with this signature, it almost looks as if the function is popping `localsCnt` times, rather than popping once and passing `localsCnt` to `finalize`. (This actually confused me for a few seconds.) Is there a better name for this that isn't too awkward? Maybe we shouldn't call the functions push/pop?

JanJecmen replied:
Yeah, sorry, I was looking at that for a while and then left it like this... Renamed the method to `finalizeCode`

mhyee commented on rir/src/compiler/translations/pir_2_rir.cpp:
```diff
@@ -0,0 +1,601 @@
+
+void Pir2Rir::removePhis(Code* code) {
+
+    // For each Phi, insert copies
```
Can you elaborate here? I'm not sure what's going on.

o- replied:
ah, yeah I was also struggling with the name here, but then forgot, after I understood what it does :)  Boissinot etal calls this "converting from SSA to CSSA (conventional ssa)"

o- replied:
btw. (but that's for later), the copies should be parallel, to avoid introducing artificial interference.

I think we should have the following:

every BB has an "out-copy" and an "in-phi" table. Then the conversion would look like this:

```
BB0:
  a0 = 0
  b0 = 0

BB1:
  goto BB3

BB2:
  a2 = 2
  b2 = 2
  goto BB3

BB3:
  x = phi(a0/BB1, b2/BB2)
  y = phi(b0/BB1, a2/BB2)
```

in this case we would say:

```
BB1-out:
    a1o, b1o = a0, b0

BB2-out:
    a2o, b20 = a2, b2

BB3-in:
    x, y = a1o | b2o , b1o | a2o
```

o- replied:
then, basically the removePhis pass would (instead of inserting copy instructions) populate the in-/out- tables

mhyee reviewed (commented):
I still need to look at `Pir2Rir::compile` in more detail, but for now, here's some questions and comments.

JanJecmen committed:
WIP PR comments + change to dt slots

JanJecmen committed:
disable p2r tests

JanJecmen commented:
I addressed the comments and started the refactoring to use dispatch table slots. It is ugly with the slots now, 0 is plain RIR, 1 is PIR optimized RIR, and the offsets are just magic numbers in the code...
Gonna continue on that.

I will squash later before merge I guess? Is that OK @o-?

JanJecmen committed:
enable p2r tests

JanJecmen committed:
fix stvar p2r

o- commented:
sorry, are you waiting for feedback on this one? it looks good to me

JanJecmen commented:
Thanks, any further comments we can fix in a new pr...
